tests/msc2716_test.go

```diff
@@ -1211,3 +1254,159 @@ func batchSendHistoricalMessages(
 
 	return res
 }
+
+// Verify that the batch of historical messages looks correct and in order in
+// the message scrollback. We can also optionally check that the historical
+// state resolves for that chunk of messages.
+//
+// Note: the historical state will only resolve correctly if the first message
+// of `/messages` is one of messages in the historical batch.
+func validateBatchSendRes(t *testing.T, c *client.CSAPI, roomID string, batchSendRes *http.Response, validateState bool) {
+	t.Helper()
+
+	batchSendResBody0 := client.ParseJSON(t, batchSendRes)
+	// Since the original body can only be read once, create a new one from the
+	// body bytes we just read
+	batchSendRes.Body = ioutil.NopCloser(bytes.NewBuffer(batchSendResBody0))
+
+	historicalEventIDs := client.GetJSONFieldStringArray(t, batchSendResBody0, "event_ids")
+	stateEventIDs := client.GetJSONFieldStringArray(t, batchSendResBody0, "state_event_ids")
+	batchEventID := client.GetJSONFieldStr(t, batchSendResBody0, "batch_event_id")
+	insertionEventID := client.GetJSONFieldStr(t, batchSendResBody0, "insertion_event_id")
+	baseInsertionEventID := gjson.GetBytes(batchSendResBody0, "base_insertion_event_id").Str
+
+	var expectedEventIDOrder []string
+	if baseInsertionEventID != "" {
+		expectedEventIDOrder = append(expectedEventIDOrder, baseInsertionEventID)
+	}
+	expectedEventIDOrder = append(expectedEventIDOrder, batchEventID)
+	expectedEventIDOrder = append(expectedEventIDOrder, reversed(historicalEventIDs)...)
+	expectedEventIDOrder = append(expectedEventIDOrder, insertionEventID)
+
+	if validateState {
+		// Get the pagination token for the end of the historical batch itself
```

Comment:
I don't like the term "end". It's so hard to know if that means old or new, especially since this entire MSC revolves around making old messages at physically later points in time. It's a tricky one to get terminology right but it would help readability on these tests:
 - start / end : ambiguous time, but clear on index positions (though it's interesting that we then access `[0]` in `:1288` :(
 - earliest / latest : ambiguous time (earliest at send time or event time?), ambiguous on index positions (depends on sort order)

You know this problem space better than I do, perhaps you can come up with better names.

Reply:
`newest-in-time` / `oldest-in-time` is probably the best for describing which end of the list we're at.

For describing the order, `chronological` / `reverse chronological` also works but it's a bit wordy and can kinda get confusing to reason about.